### ISSUE TITLE
Structure of Checkpoint file

### ISSUE THREAD

**Issue by ayush1999** [keras-team/keras#10026]
I was working on a project to make a reader for Keras models for another framework written in Julia. The model will be read from the Keras checkpoint file. Can someone help me understand the structure of this checkpoint file? As in how is the data about the layers, parameters stored here? Any other resources which may be helpful?
Thanks a lot.

**Comment by TheShip1948** [comment]
Hi @ayush1999 , 

**Hint:** all links doesn't work on clicking on them, you need to copy and best the link (not sure why). 

Checkpoint file is saved in HDF5 format. I couldn't find online documentation to understand the content of the file. The following two ways are the ones that will help you to have better understanding to the content of the checkpoint file content. 

**First:** To be able to understand its content, you need to view the file content with HDF5 viewer. Fortunately, you could find some viewers you can download. The following is one of the viewers that you can use. 
HDF5 Viewer Download Link: [https://support.hdfgroup.org/products/java/release/download.html](url) 
 
**Other way:** to understand the content of the file is to read the implementation of the ModelCheckPoint function that writes the checkpoint file. This function calls save_model function that do the actual saving. 
**ModelCheckPoint file:** [https://github.com/keras-team/keras/blob/master/keras/callbacks.py](url) 
**Save_model file:** [https://github.com/keras-team/keras/blob/master/keras/engine/saving.py](url)

**Tutorial:** 
This is a useful tutorial for checkpoint generation and loading. 
Link: [https://machinelearningmastery.com/check-point-deep-learning-models-keras/](url)

You can find a link to photo that views the content of the result of a code in the tutorial in the following link. 
[https://1drv.ms/u/s!AmC0MkRjoIu2gQiDFSE_IF4njppS](url) 

Snippet of the representation for the model used in the tutorial: 
   Group size = 2
    Number of attributes = 4
        backend = tensorflow
        keras_version = 2.1.6
        model_config = {"class_name": "Sequential", "config": [{"class_name": "Dense", "config": {"kernel_initializer": {"class_name": "RandomUniform", "config": {"maxval": 0.05, "seed": null, "minval": -0.05}}, "name": "dense_1", "kernel_constraint": null, "bias_regularizer": null, "bias_constraint": null, "dtype": "float32", "activation": "relu", "trainable": true, "kernel_regularizer": null, "bias_initializer": {"class_name": "Zeros", "config": {}}, "units": 12, "batch_input_shape": [null, 8], "use_bias": true, "activity_regularizer": null}}, {"class_name": "Dense", "config": {"kernel_initializer": {"class_name": "RandomUniform", "config": {"maxval": 0.05, "seed": null, "minval": -0.05}}, "name": "dense_2", "kernel_constraint": null, "bias_regularizer": null, "bias_constraint": null, "activation": "relu", "trainable": true, "kernel_regularizer": null, "bias_initializer": {"class_name": "Zeros", "config": {}}, "units": 8, "use_bias": true, "activity_regularizer": null}}, {"class_name": "Dense", "config": {"kernel_initializer": {"class_name": "RandomUniform", "config": {"maxval": 0.05, "seed": null, "minval": -0.05}}, "name": "dense_3", "kernel_constraint": null, "bias_regularizer": null, "bias_constraint": null, "activation": "sigmoid", "trainable": true, "kernel_regularizer": null, "bias_initializer": {"class_name": "Zeros", "config": {}}, "units": 1, "use_bias": true, "activity_regularizer": null}}]}
        training_config = {"metrics": ["accuracy"], "loss": "binary_crossentropy", "optimizer_config": {"class_name": "Adam", "config": {"beta_1": 0.8999999761581421, "beta_2": 0.9990000128746033, "decay": 0.0, "epsilon": 1e-07, "amsgrad": false, "lr": 0.0010000000474974513}}, "loss_weights": null, "sample_weight_mode": null}

you can see this is the format used inside the file. Off course you will need HDF5 APIs to parse the file and filter the content that you want from it. 

I hope this info is enough, and please feel free to ask if anything isn't clear.

**Comment by ayush1999** [comment]
@TheShip1948 Sorry, was a bit busy. I needed help with two things:
1. Given the checkpoint file, I want to extract the weights and store them in a dictionary
2. Given the file, I need to get the structure stores, so as to read it in another framework.

What is your take on this?

**Comment by TheShip1948** [comment]
Hi @ayush1999  
I think all you need as I mentioned in previous post is HDF5 reader in Julia. This will be able to read the file. As the image I mentioned in the past, the reader. Then use the reader to query the information you need. The Image I mentioned in the past reply  (https://1drv.ms/u/s!AmC0MkRjoIu2gQiDFSE_IF4njppS) shows how the reader can show the content, so that you can easily query for the information you need form the file.

**Comment by ayush1999** [comment]
Hey @TheShip1948 . I was using the Julia HDF5 reader (HDF5.jl). After reading a sample (https://github.com/EN10/KerasMNIST/blob/master/TFKeras.h5) using this reader, here is what I obtained:

The data contains a dict, which contains two keys: `"model_weights"` and `"optimizer_weights"`. Each of these keys contain a tensor of numbers, which are the weights of the model. What I'm not able to infer from this data is the structure of the model isn't visible here. So, if I had to build this model in another framework, I'd need three things:

1. The structure of the model from the .h5 file.
2. The weights accompanying each layer.

Can you please help me out here? Thanks a for for your time.
